### DISCUSSION TITLE
How to generate a PDF and download it without rendering it?

### DISCUSSION THREAD

**Discussion by geeky01adarsh** [diegomura/react-pdf#2352] [Q&A] [answered]
I was trying to generate a PDF with some layout. I have configured the values using hooks, but I want to download the generated PDF without rendering it to the client side. How can I achieve this with React?

**Comment by CoskunCakir** [comment, marked as answer]
hey @geeky01adarsh, I believe you just don't want to display it right? In that case, you could use the `@react-pdf/renderer`s pdf method and the `file-saver` package to achieve this.

```jsx
import React from 'react';
import { saveAs } from 'file-saver';
import { pdf } from '@react-pdf/renderer';
import YourDocument from './YourDocument';

const DownloadButton = () => {
  const downloadPdf = async () => {
    const fileName = 'test.pdf';
    const blob = await pdf(<YourDocument />).toBlob();
    saveAs(blob, fileName);
  };

  return <button onClick={downloadPdf}>Download PDF</button>;
};

export default DownloadButton;
```

this will convert your pdf component into a Blob and then saves it using saveAs method from file-saver

**Reply by geeky01adarsh** [reply]
That's what I exactly wanted, thanks for the help @coskuncakir.

**Reply by Venkata2138279** [reply]
is there a way to add password protection to the PDF

**Reply by Qodestackr** [reply]
> is there a way to add password protection to the PDF

Hello, any way?

**Reply by HetarthJain** [reply]
Hi, @coskuncakir my pdf is very large, >50 pages. how to use *workers* in ReactJS to generate the pdf (using React-pdf itself) using same method as given above? 
I noticed that, this line of code blocks the main browser thread.
``` const blob = await pdf(<YourDocument />).toBlob(); ```
i tried,
``` import { pdf } from '@react-pdf/renderer; ``` 
I was trying to import pdf inside app.worker.js file,
passing ```<YourDocument>``` from app.js file to worker file by postMessage.
I read that we can pass only serializable data to workers, and we cannot import modules inside workers. 
How to approach this? Please help here!

**Reply by SameedRazi** [reply]
> Hi, @CoskunCakir my pdf is very large, >50 pages. how to use _workers_ in ReactJS to generate the pdf (using React-pdf itself) using same method as given above? I noticed that, this line of code blocks the main browser thread. `const blob = await pdf(<YourDocument />).toBlob();` i tried, `import { pdf } from '@react-pdf/renderer;` I was trying to import pdf inside app.worker.js file, passing `<YourDocument>` from app.js file to worker file by postMessage. I read that we can pass only serializable data to workers, and we cannot import modules inside workers. How to approach this? Please help here!

were you able to fix this issue?

**Reply by HetarthJain** [reply]
No. The jsx is not serializable to be converted into blob

**Comment by Qodestackr** [comment]
Can I also like Generate PDF client side while triggering a server action to get the link of a resource? (Nextjs that is).
